File: scripts/federated_campaign/plot_fixed_stage_latest.R
#!/usr/bin/env Rscript

args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 10) {
  stop(paste(
    "usage: plot_fixed_stage_latest.R <csv> <metric> <ylabel> <title>",
    "<subtitle> <footer> <column-order-csv> <total-cells> <png> <svg>"
  ))
}

csv_path <- args[[1]]
metric <- args[[2]]
y_label <- args[[3]]
plot_title <- args[[4]]
plot_subtitle <- args[[5]]
plot_footer <- args[[6]]
workloads <- strsplit(args[[7]], ",", fixed = TRUE)[[1]]
total_cells <- as.integer(args[[8]])
png_path <- args[[9]]
svg_path <- args[[10]]

data <- read.csv(csv_path, stringsAsFactors = FALSE)
profiles <- c("lan", "wan_light", "wan_mid")
profile_labels <- c(lan = "LAN", wan_light = "WAN-Light", wan_mid = "WAN-Mid")
planners <- c("FedAll", "Heuristic", "DP", "Exact")
legend_labels <- c(FedAll = "ALL", Heuristic = "Heuristic", DP = "DP", Exact = "Exact")
colors <- c(FedAll = "#c44e52", Heuristic = "#55a868", DP = "#e17c05", Exact = "#4c72b0")
symbols <- c(FedAll = 0, Heuristic = 1, DP = 5, Exact = 2)
label_positions <- c(FedAll = 3, Heuristic = 1, DP = 3, Exact = 1)

if (!(metric %in% names(data))) {
  stop(paste("missing metric column:", metric))
}
if (length(workloads) != 7 || length(unique(workloads)) != 7) {
  stop("expected seven unique workloads")
}
if (nrow(data) == 0 || nrow(data) > total_cells) {
  stop("invalid successful-row count")
}
data[[metric]] <- as.numeric(data[[metric]])
if (any(!is.finite(data[[metric]])) || any(data[[metric]] < 0)) {
  stop("metric contains invalid values")
}

column_ymax <- setNames(rep(1, length(workloads)), workloads)
for (workload in workloads) {
  values <- data[data$workload == workload, metric]
  observed <- if (length(values) > 0) max(values) else 0
  column_ymax[[workload]] <- if (observed > 0) observed * 1.16 else 1
}

draw_graph <- function() {
  panel_ids <- matrix(2:22, nrow = length(profiles), byrow = TRUE)
  layout(
    rbind(rep(1, length(workloads)), panel_ids),
    heights = c(0.17, rep(0.2767, length(profiles)))
  )
  par(
    oma = c(1.25, 4.0, 0.2, 0.45),
    mgp = c(2.0, 0.58, 0),
    tcl = -0.24,
    las = 1,
    family = "sans"
  )

  par(mar = c(0, 0, 0, 0))
  plot.new()
  text(0.5, 0.84, plot_title, cex = 1.16, font = 2)
  text(0.5, 0.58, plot_subtitle, cex = 0.69, col = "#5f5f5f")
  legend(
    x = 0.5,
    y = 0.25,
    legend = legend_labels[planners],
    col = colors[planners],
    pch = symbols[planners],
    lty = 1,
    lwd = 1.55,
    horiz = TRUE,
    bty = "n",
    cex = 0.86,
    xjust = 0.5,
    yjust = 0.5
  )
  text(0.5, 0.03, plot_footer, cex = 0.56, col = "#777777")

  for (profile_index in seq_along(profiles)) {
    profile <- profiles[[profile_index]]
    for (workload_index in seq_along(workloads)) {
      workload <- workloads[[workload_index]]
      block <- data[data$profile == profile & data$workload == workload, ]
      bottom_row <- profile_index == length(profiles)
      ymax <- column_ymax[[workload]]
      par(mar = c(if (bottom_row) 3.45 else 2.05, 2.3, 2.0, 0.48))

      plot(
        NA,
        xlim = c(0.78, 4.22),
        ylim = c(0, ymax),
        xaxt = "n",
        xlab = if (bottom_row) "Workers" else "",
        ylab = "",
        main = if (profile_index == 1) toupper(workload) else "",
        cex.main = 0.98,
        cex.lab = 0.82,
        cex.axis = 0.70,
        bty = "l"
      )
      axis(1, at = 1:4, labels = 1:4, cex.axis = 0.70)
      abline(h = pretty(c(0, ymax)), col = "#dddddd", lty = 2, lwd = 0.72)

      if (workload_index == 1) {
        mtext(profile_labels[[profile]], side = 2, line = 3.0, cex = 0.84, las = 0)
      }

      for (planner in planners) {
        values <- rep(NA_real_, 4)
        series <- block[block$planner == planner, c("workers", metric)]
        if (nrow(series) > 0) {
          values[as.integer(series$workers)] <- as.numeric(series[[metric]])
        }
        lines(
          1:4,
          values,
          type = "o",
          col = colors[[planner]],
          pch = symbols[[planner]],
          lwd = 1.55,
          cex = 0.70
        )
        observed <- which(is.finite(values))
        if (length(observed) > 0) {
          text(
            observed,
            values[observed],
            labels = sprintf("%.2f", values[observed]),
            pos = label_positions[[planner]],
            offset = 0.22,
            cex = 0.43,
            col = colors[[planner]],
            xpd = FALSE
          )
        }
      }

      count <- nrow(block)
      if (count == 0) {
        text(2.5, ymax * 0.50, "not yet executed", cex = 0.70, col = "#999999")
      } else if (count < 16) {
        text(4.12, ymax * 0.96, paste0("partial ", count, "/16"),
             cex = 0.60, col = "#777777", adj = c(1, 1))
      }
    }
  }

  mtext(y_label, outer = TRUE, side = 2, line = 1.60, cex = 0.91, las = 0)
}

dir.create(dirname(png_path), recursive = TRUE, showWarnings = FALSE)
png(png_path, width = 4200, height = 1800, res = 200, pointsize = 12)
draw_graph()
dev.off()

svg(svg_path, width = 21, height = 9, pointsize = 12)
draw_graph()
dev.off()

File: scripts/federated_campaign/plot_fixed_stage_latest.py
#!/usr/bin/env python3
"""Render authenticated partial-campaign runtime and compile-time grids."""

from __future__ import annotations

import argparse
from collections import Counter
import csv
from datetime import datetime
import hashlib
import json
import math
from pathlib import Path
import re
import shutil
import subprocess
import time
from typing import Any


PROFILES = ("lan", "wan_light", "wan_mid")
WORKLOADS = ("pca", "lm", "kmeans", "l2svm", "logreg", "als", "steplm")
PLANNERS = ("FedAll", "Heuristic", "DP", "Exact")
WORKERS = (1, 2, 3, 4)
COMPILATION_PATTERN = re.compile(
    r"^Total compilation time:\s*([0-9]+(?:\.[0-9]+)?) sec\.$", re.MULTILINE
)


def sha256_file(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for block in iter(lambda: handle.read(1024 * 1024), b""):
            digest.update(block)
    return digest.hexdigest()


def load_json(path: Path) -> dict[str, Any]:
    return json.loads(path.read_text(encoding="utf-8"))


def validate_row(row: dict[str, Any], manifest_hash: str) -> dict[str, Any]:
    profile = str(row["profile"])
    workload = str(row["workload"])
    planner = str(row["planner"])
    workers = int(row["workers"])
    if profile not in PROFILES or workload not in WORKLOADS:
        raise RuntimeError("unexpected profile/workload")
    if planner not in PLANNERS or workers not in WORKERS:
        raise RuntimeError("unexpected planner/workers")
    cell = f"workers={workers}|planner={planner}|workload={workload}|profile={profile}"
    if row.get("cell") != cell or row.get("manifest_sha256") != manifest_hash:
        raise RuntimeError("cell identity or manifest link mismatch")
    trace_counts = row.get("planning_evidence", {}).get("planner_trace_stage_counts", {})
    if not (
        row.get("oracle_passed") is True
        and row.get("runtime_scan_clean") is True
        and row.get("teardown_zero_resources") is True
        and row.get("fallback") is False
        and row.get("primary_metric_phase") == "warm-fresh-coordinator-jvm"
        and row.get("runtime_plan_sha256") == row.get("cold_runtime_plan_sha256")
        and row.get("instruction_fingerprint") == row.get("cold_instruction_fingerprint")
        and trace_counts.get("Planner-Invoke") == 1
        and trace_counts.get("Planner-Complete") == 1
        and trace_counts.get("Emission-Summary") == 1
    ):
        raise RuntimeError("authenticated successful-row contract failed")

    execution_seconds = float(row["execution_seconds"])
    warm_seconds = float(row["warm_evidence"]["seconds"])
    if not math.isfinite(execution_seconds) or execution_seconds <= 0:
        raise RuntimeError("invalid execution time")
    if not math.isclose(execution_seconds, warm_seconds, rel_tol=0, abs_tol=1e-9):
        raise RuntimeError("warm execution metric mismatch")

    warm_log = Path(row["warm_bundle"]) / "raw_coordinator.log"
    if not warm_log.is_file():
        raise RuntimeError("missing warm raw coordinator log")
    warm_hash = sha256_file(warm_log)
    if warm_hash != row["warm_evidence"]["raw_coordinator_sha256"]:
        raise RuntimeError("warm raw coordinator hash mismatch")
    matches = COMPILATION_PATTERN.findall(
        warm_log.read_text(encoding="utf-8", errors="replace")
    )
    if len(matches) != 1:
        raise RuntimeError(f"expected one compilation metric, found {len(matches)}")
    compilation_seconds = float(matches[0])
    if not math.isfinite(compilation_seconds) or compilation_seconds < 0:
        raise RuntimeError("invalid compilation time")

    return {
        "profile": profile,
        "workload": workload,
        "workers": workers,
        "planner": planner,
        "execution_seconds": execution_seconds,
        "total_compilation_seconds": compilation_seconds,
        "attempt": int(row.get("attempt", 0)),
        "cell": cell,
        "warm_bundle": str(row["warm_bundle"]),
        "warm_raw_coordinator_sha256": warm_hash,
        "runtime_plan_sha256": str(row["runtime_plan_sha256"]),
    }


def read_stable_rows(root: Path) -> tuple[list[dict[str, Any]], dict[str, Any], dict[str, Any]]:
    manifest = load_json(root / "manifest.json")
    manifest_hash = str(manifest["manifest_sha256"])
    rows_path = root / "rows.jsonl"
    last_error = "campaign did not become snapshot-stable"
    for _ in range(40):
        before = load_json(root / "progress.json")
        raw_rows = []
        try:
            with rows_path.open(encoding="utf-8") as handle:
                for line_number, line in enumerate(handle, start=1):
                    if line.strip():
                        raw_rows.append((line_number, json.loads(line)))
        except json.JSONDecodeError as error:
            last_error = f"rows.jsonl changed during read: {error}"
            time.sleep(0.25)
            continue
        after = load_json(root / "progress.json")
        if before != after or before.get("completed") != len(raw_rows):
            last_error = (
                f"unstable progress/row count: before={before.get('completed')} "
                f"rows={len(raw_rows)} after={after.get('completed')}"
            )
            time.sleep(0.25)
            continue

        selected: dict[str, tuple[int, dict[str, Any]]] = {}
        for line_number, raw in raw_rows:
            validated = validate_row(raw, manifest_hash)
            prior = selected.get(validated["cell"])
            rank = (validated["attempt"], line_number)
            if prior is None or rank > (prior[1]["attempt"], prior[0]):
                selected[validated["cell"]] = (line_number, validated)
        rows = [item[1] for item in selected.values()]
        if len(rows) != int(before["completed"]):
            raise RuntimeError("duplicate canonical successful cells in immutable campaign")
        return sorted(rows, key=lambda row: row["cell"]), before, manifest
    raise RuntimeError(last_error)


def runtime_separation(rows: list[dict[str, Any]], workload: str) -> float:
    spreads = []
    for profile in PROFILES:
        for workers in WORKERS:
            values = {
                row["planner"]: float(row["execution_seconds"])
                for row in rows
                if row["profile"] == profile
                and row["workload"] == workload
                and row["workers"] == workers
            }
            if set(values) == set(PLANNERS):
                spreads.append(max(values.values()) - min(values.values()))
    if not spreads:
        return math.inf
    return max(spreads)


def write_snapshot(path: Path, rows: list[dict[str, Any]]) -> None:
    fields = (
        "profile",
        "workload",
        "workers",
        "planner",
        "execution_seconds",
        "total_compilation_seconds",
        "attempt",
        "cell",
        "warm_bundle",
        "warm_raw_coordinator_sha256",
        "runtime_plan_sha256",
    )
    with path.open("w", encoding="utf-8", newline="") as handle:
        writer = csv.DictWriter(handle, fieldnames=fields)
        writer.writeheader()
        writer.writerows(rows)


def render(
    plotter: Path,
    snapshot: Path,
    metric: str,
    ylabel: str,
    title: str,
    subtitle: str,
    footer: str,
    column_order: list[str],
    total: int,
    png: Path,
    svg: Path,
) -> None:
    command = [
        "Rscript",
        str(plotter),
        str(snapshot),
        metric,
        ylabel,
        title,
        subtitle,
        footer,
        ",".join(column_order),
        str(total),
        str(png),
        str(svg),
    ]
    result = subprocess.run(command, text=True, capture_output=True, check=False)
    if result.returncode != 0:
        raise RuntimeError(f"R plot failed:\n{result.stdout}\n{result.stderr}")
    if not png.is_file() or not svg.is_file():
        raise RuntimeError("plot command returned without both output files")


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("result_root", type=Path)
    args = parser.parse_args()
    root = args.result_root.resolve()
    rows, progress, manifest = read_stable_rows(root)
    completed = len(rows)
    total = int(progress["total"])
    plots = root / "plots"
    plots.mkdir(parents=True, exist_ok=True)
    snapshot = plots / f"fixed_stage_snapshot_{completed}of{total}.csv"
    write_snapshot(snapshot, rows)

    canonical_rank = {workload: index for index, workload in enumerate(WORKLOADS)}
    separation = {workload: runtime_separation(rows, workload) for workload in WORKLOADS}
    column_order = sorted(
        WORKLOADS,
        key=lambda workload: (separation[workload], canonical_rank[workload]),
    )
    generated_at = datetime.now().astimezone().isoformat(timespec="seconds")
    subtitle = (
        "Rows: LAN -> WAN-Light -> WAN-Mid; columns: increasing max runtime spread "
        "over complete 4-planner blocks; missing cells are not interpolated"
    )
    footer = (
        f"source {str(manifest['systemds_commit'])[:7]} | harness "
        f"{str(manifest['harness_commit'])[:7]} | fixed seed {manifest['campaign_seed']} | "
        "warm fresh-coordinator JVM"
    )
    plotter = Path(__file__).with_suffix(".R")
    prefix = f"{completed}of{total}"
    outputs = {
        "runtime_png": plots / f"runtime_grid_3x7_fixed_stage_{prefix}.png",
        "runtime_svg": plots / f"runtime_grid_3x7_fixed_stage_{prefix}.svg",
        "compile_png": plots / f"compile_time_grid_3x7_fixed_stage_{prefix}.png",
        "compile_svg": plots / f"compile_time_grid_3x7_fixed_stage_{prefix}.svg",
    }
    render(
        plotter,
        snapshot,
        "execution_seconds",
        "Execution time (seconds)",
        f"Fixed-stage Docker runtime — {completed}/{total} completed cells",
        subtitle,
        footer,
        column_order,
        total,
        outputs["runtime_png"],
        outputs["runtime_svg"],
    )
    render(
        plotter,
        snapshot,
        "total_compilation_seconds",
        "Compilation time (seconds)",
        f"Fixed-stage Docker compilation — {completed}/{total} completed cells",
        subtitle,
        footer,
        column_order,
        total,
        outputs["compile_png"],
        outputs["compile_svg"],
    )

    latest = {
        "runtime_latest_png": plots / "runtime_grid_3x7_fixed_stage_latest.png",
        "runtime_latest_svg": plots / "runtime_grid_3x7_fixed_stage_latest.svg",
        "compile_latest_png": plots / "compile_time_grid_3x7_fixed_stage_latest.png",
        "compile_latest_svg": plots / "compile_time_grid_3x7_fixed_stage_latest.svg",
    }
    shutil.copyfile(outputs["runtime_png"], latest["runtime_latest_png"])
    shutil.copyfile(outputs["runtime_svg"], latest["runtime_latest_svg"])
    shutil.copyfile(outputs["compile_png"], latest["compile_latest_png"])
    shutil.copyfile(outputs["compile_svg"], latest["compile_latest_svg"])

    profile_counts = Counter(row["profile"] for row in rows)
    planner_counts = Counter(row["planner"] for row in rows)
    workload_counts = Counter(row["workload"] for row in rows)
    paths = {**outputs, **latest, "snapshot_csv": snapshot}
    receipt = {
        "schema": "g014-fixed-stage-plot/v2",
        "generated_at": generated_at,
        "active_output": str(root),
        "stage_id": manifest["stage_id"],
        "systemds_commit": manifest["systemds_commit"],
        "harness_commit": manifest["harness_commit"],
        "systemds_jar_sha256": manifest["systemds_jar_sha256"],
        "manifest_sha256": manifest["manifest_sha256"],
        "progress_snapshot": progress,
        "row_count": completed,
        "profile_counts": dict(sorted(profile_counts.items())),
        "planner_counts": dict(sorted(planner_counts.items())),
        "workload_counts": dict(sorted(workload_counts.items())),
        "column_order": column_order,
        "column_runtime_separation_seconds": {
            workload: None if math.isinf(separation[workload]) else separation[workload]
            for workload in column_order
        },
        "runtime_metric": "execution_seconds from warm-fresh-coordinator-jvm",
        "compile_metric": "Total compilation time from the same warm raw coordinator log",
        "selection_policy": (
            "latest authenticated successful canonical row from the current immutable "
            "stage only; snapshot progress and rows must be stable and equal"
        ),
        "separation_policy": (
            "maximum runtime max-minus-min across profile/worker blocks containing all "
            "four planners; ascending; canonical workload order breaks ties"
        ),
        "y_axis_policy": (
            "zero lower bound; common maximum per workload column across environment rows"
        ),
        "outputs": {name: str(path) for name, path in paths.items()},
        "sha256": {name: sha256_file(path) for name, path in paths.items()},
    }
    receipt_path = plots / "latest_fixed_stage_plot_receipt.json"
    receipt_path.write_text(
        json.dumps(receipt, indent=2, sort_keys=True) + "\n", encoding="utf-8"
    )
    print(json.dumps(receipt, indent=2, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
